### PR TITLE
Misc DFP Documentation & fixes

### LINKED ISSUE #345: [DOC] Complete Documentation for the Production DFP Workflow Example
## Report needed documentation

**Report needed documentation**
The Production DFP Workflow example needs to be documented to make it easy for users to get started.

**Describe the documentation you'd like**
The documentation should include:

- [x] A statement of the problem that is solved with this workflow
   - [x] Including an example of the input data
- [x] How the Production DFP example differs from the Production
   - [x] Including all of the various services
- [ ] How to run the example:
   - [x] Via `docker-compose`
   - [ ] Via the helm chart
- [x] Where to extend this example for different input datasets

### PR DESCRIPTION
**Pull request by dagardner-nv**
* Fills in `examples/digital_fingerprinting/production/README.md` with a bit more detail
* Fix handling of command line enum values
* Add date window filtering to `DFPFileBatcherStage`, since this stage is already performing date calculations, and occurs prior to the downloading of any remote data. Previously this was a part of the `S3SourceStage`.
* Add a `--start_time` flag to `dfp_azure_pipeline.py` and `dfp_duo_pipeline.py` allowing an explicit time window.
* `morpheus_training` service in docker compose renamed to `morpheus_pipeline` to reflect that it is used for both training and inference
* Set nvidia runtime in docker compose yaml for users who don't have it as their default runtime.
* MLflow version restricted to <1.29 to avoid a bug (per @pdmack )
* Fix casing of class names in headings for Sphinx builds
* Spelling fixes in docs & help strings
* Fix launch script for `morpheus_pipeline`
* from-azure and from-duo added to CLI
* Remove usage of `only_new_batches` argument which no longer exists
* CI work-around for https://gitlab.com/karolinepauls/pytest-kafka/-/issues/10
* Update diagrams to reflect recent code changes
* Support `_` as a time separator in `iso_date_regex`
* Adds the following to `docs/source/developer_guide/guides/5_digital_fingerprinting.md`:
  * Section on how to define a new data source schema
  * Explanation of Starter & Production examples
  * docker-compose commands for building & starting services in the production example
  * Helm Chart info for production example
  * Fix headings for stage classes

Fixes #345

### CODE REVIEW

File: docs/source/developer_guide/guides/5_digital_fingerprinting.md [hunk 1]
```diff
@@ -23,12 +23,54 @@ Every account, user, service and machine has a digital fingerprint​, which rep
 To construct this digital fingerprint we will be training unsupervised behavioral models at various granularities, including a generic model for all users in the organization along with fine-grained models for each user to monitor their behavior. These models are continuously updated and retrained overtime​, and alerts are triggered when deviations from normality occur for any user​.
```
Comment (efajardo-nv) [inline docs/source/developer_guide/guides/5_digital_fingerprinting.md:23]: overtime -> over time

File: docs/source/developer_guide/guides/5_digital_fingerprinting.md [hunk 2]
```diff
@@ -23,12 +23,54 @@ Every account, user, service and machine has a digital fingerprint​, which rep
...
+The location of these logs could be either local to the machine running Morpheus, a shared file system like NFS or on a remote store such as [Amazon S3](https://aws.amazon.com/s3/).
 
-Additional data sources and remote stores can easily be added using the Morpheus SDK, the key to applying DFP to a new data source is through the process of feature selection. Any data source can be fed into DFP after some preprocessing to get a feature vector per log/data point​.  Since DFP builds a targeted model for each entity (user/service/machine... etc.), it would work best if the chosen data source has a field that uniquely identifies the entity we’re trying to model.
+### Defining a New Data Source
+Additional data sources and remote stores can easily be added using the Morpheus SDK, the key to applying DFP to a new data source is through the process of feature selection. Any data source can be fed into DFP after some preprocessing to get a feature vector per log/data point​.  In order to build a targeted model for each entity (user/service/machine... etc.), the chosen data source needs a field that uniquely identifies the entity we’re trying to model.
```
Comment (efajardo-nv) [inline docs/source/developer_guide/guides/5_digital_fingerprinting.md:31]: Split first sentence into two sentences.

File: docs/source/developer_guide/guides/5_digital_fingerprinting.md [hunk 3]
```diff
@@ -167,33 +280,152 @@ Other attributes which might be needed:
...
+
+### Schema Definition
+
+#### DataFrame Input Schema (`DataFrameInputSchema`)
+The `DataFrameInputSchema` ([examples/digital_fingerprinting/production/morpheus/dfp/utils/column_info.py](/examples/digital_fingerprinting/production/morpheus/dfp/utils/column_info.py)) class defines the schema specifying the columns to be included in the output `DataFrame`.  Within the DFP pipeline there are two stages where pre-processing is performed, the `DFPFileToDataFrameStage` stage and the `DFPPreprocessingStage`.  This decoupling of the pre-processing stages from the actual operations needed to be performed allows for the actual schema to be user-defined in the pipeline and re-usability of the stages.  It is up to the user to define the fields which will appear in the `DataFrame`, any column in the input data that isn't specified in either `column_info` or `preserve_columns` constructor arguments will not appear in the output.   The exception to this are JSON fields, specified in the `json_columns` argument which defines json fields which are to be normalized.
```
Comment (efajardo-nv) [inline docs/source/developer_guide/guides/5_digital_fingerprinting.md:288]: split into two sentences:
```
It is up to the user to define the fields which will appear in the `DataFrame`. Any column in the input data that isn't specified in either `column_info` or `preserve_columns` constructor arguments will not appear in the output.
```

File: docs/source/developer_guide/guides/5_digital_fingerprinting.md [hunk 4]
```diff
@@ -210,15 +442,15 @@ pipeline.add_stage(
...
+#### File to DataFrame Stage (`DFPFileToDataFrameStage`)
+The `DFPFileToDataFrameStage` ([examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_file_to_df.py](/examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_file_to_df.py)) stage receives a `list` of an [fsspec.core.OpenFiles](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.core.OpenFiles) and loads them into a single `DataFrame` which is then emitted into the pipeline.  When the parent stage is `DFPFileBatcherStage` each batch (typically one day) is concatenated into a single `DataFrame`, without this if the parent was `MultiFileSource` the entire dataset is loaded into a single `DataFrame`.  Because of this, it is important to choose a `period` argument for `DFPFileBatcherStage` small enough such that each batch can fit into memory.
```
Comment (efajardo-nv) [inline docs/source/developer_guide/guides/5_digital_fingerprinting.md:448]: Split into two sentences:
```
When the parent stage is DFPFileBatcherStage each batch (typically one day) is concatenated into a single DataFrame. If the parent was MultiFileSource the entire dataset is loaded into a single DataFrame. 
```

Comment (efajardo-nv) [inline docs/source/developer_guide/guides/5_digital_fingerprinting.md:448]: Split into two sentences:
```
This stage is able to download and load data files concurrently by multiple methods. Currently supported methods are: single_thread, multiprocess, dask, and dask_thread. 
```

File: docs/source/developer_guide/guides/5_digital_fingerprinting.md [hunk 5]
```diff
@@ -228,58 +460,73 @@ This stage is able to download & load data files concurrently by multiple method
...
+These stages are common to both the training & inference pipelines, unlike the input & output stages these are specific to the DFP pipeline and intended to be configured but not replaceable.
 
-#### DFPSplitUsersStage
-The `DFPSplitUsersStage` [examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_split_users_stage.py](/examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_split_users_stage.py) stage recieves an incoming `DataFrame` and emits a `list` of `DFPMessageMeta` where each `DFPMessageMeta` represents the records associated for a given user.  This allows for downstream stages to perform all nescesary opperations on a per user basis.
+#### Split Users Stage (`DFPSplitUsersStage`)
+The `DFPSplitUsersStage` [examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_split_users_stage.py](/examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_split_users_stage.py) stage receives an incoming `DataFrame` and emits a `list` of `DFPMessageMeta` where each `DFPMessageMeta` represents the records associated for a given user.  This allows for downstream stages to perform all necessary operations on a per user basis.
```
Comment (efajardo-nv) [inline docs/source/developer_guide/guides/5_digital_fingerprinting.md:491]: Add paranthesis to link to be consistent with other links.

File: docs/source/developer_guide/guides/5_digital_fingerprinting.md [hunk 6]
```diff
@@ -228,58 +460,73 @@ This stage is able to download & load data files concurrently by multiple method
...
-The `DFPRollingWindowStage` [examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_rolling_window_stage.py](/examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_rolling_window_stage.py) stage will trim incoming events within a specific size or time window as specified by the `min_history`, `min_increment` and `max_history` constructor arguments.
-
-| Argument | Type | Descirption |
+| `skip_users` | `List[str]` or `None` | List of users to exclude, when `include_generic` is `True` excluded records will also be excluded from the generic user. Mutually exclusive with `only_users`.  |
+| `only_users` | `List[str]` or `None` | Limit records to a specific list of users, when `include_generic` is `True` the generic user's records will also be limited to the users in this list. Mutually exclusive with `skip_users`. |
+
+#### Rolling Window Stage (`DFPRollingWindowStage`)
+The `DFPRollingWindowStage` [examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_rolling_window_stage.py](/examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_rolling_window_stage.py) stage performs several key pieces of functionality for DFP.
```
Comment (efajardo-nv) [inline docs/source/developer_guide/guides/5_digital_fingerprinting.md:502]: Add paranthesis to link to be consistent with other links.

File: docs/source/developer_guide/guides/5_digital_fingerprinting.md [hunk 7]
```diff
@@ -228,58 +460,73 @@ This stage is able to download & load data files concurrently by multiple method
...
+#### Preprocessing Stage (`DFPPreprocessingStage`)
+The `DFPPreprocessingStage` [examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_preprocessing_stage.py](/examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_preprocessing_stage.py) stage, the actual logic of preprocessing is defined in the `input_schema` argument.  Since this stage occurs in the pipeline after the `DFPFileBatcherStage` and `DFPSplitUsersStage` stages all records in the incoming `DataFrame` correspond to only a single user within a specific time period allowing for columns to be computer on a per-user per-time period basis such as the `logcount` and `locincrement` features mentioned above.  Making the type of processing performed in this stage different from those performed in the `DFPFileToDataFrameStage`.
```
Comment (efajardo-nv) [inline docs/source/developer_guide/guides/5_digital_fingerprinting.md:524]: Add paranthesis to link to be consistent with other links.

File: docs/source/developer_guide/guides/5_digital_fingerprinting.md [hunk 8]
```diff
@@ -290,48 +537,54 @@ After training the generic model, individual user models can be trained​.  Ind
 
 ### Training Stages
 
-#### DFPTraining
+#### Training Stage (`DFPTraining`)
 The `DFPTraining` [examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_training.py](/examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_training.py) trains a model for each incoming `DataFrame` and emits an instance of `morpheus.messages.multi_ae_message.MultiAEMessage` containing the trained model.
```
Comment (efajardo-nv) [inline docs/source/developer_guide/guides/5_digital_fingerprinting.md:541]: Add paranthesis to link to be consistent with other links or make class name a link.

File: docs/source/developer_guide/guides/5_digital_fingerprinting.md [hunk 9]
```diff
@@ -290,48 +537,54 @@ After training the generic model, individual user models can be trained​.  Ind
...
 
-#### DFPMLFlowModelWriterStage
+| Argument | Type | Description |
+| -------- | ---- | ----------- |
+| `c` | `morpheus.config.Config` | Morpheus config object |
+| `model_kwargs` | `dict` or `None` | Optional dictionary of keyword arguments to be used when constructing the model.  See [`AutoEncoder`](https://github.com/nv-morpheus/dfencoder/blob/master/dfencoder/autoencoder.py) for information on the available options.|
+
+#### MLflow Model Writer Stage (`DFPMLFlowModelWriterStage`)
 The `DFPMLFlowModelWriterStage` [examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_mlflow_model_writer.py](/examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_mlflow_model_writer.py) stage publishes trained models into MLflow, skipping any model which lacked sufficient training data (current required minimum is 300 log records).
```
Comment (efajardo-nv) [inline docs/source/developer_guide/guides/5_digital_fingerprinting.md:549]: Add paranthesis to link to be consistent with other links or make class name a link.

File: docs/source/developer_guide/guides/5_digital_fingerprinting.md [hunk 10]
```diff
@@ -290,48 +537,54 @@ After training the generic model, individual user models can be trained​.  Ind
...
+> **Note:**  If using a remote MLflow server, users will need to call [`mlflow.set_tracking_uri`](https://www.mlflow.org/docs/latest/python_api/mlflow.html#mlflow.set_tracking_uri) before starting the pipeline.
 
 ## Inference Pipeline
 ![Inference Pipeline Overview](img/dfp_inference_overview.png)
 
 ### Inference Stages
 
-#### DFPInferenceStage
+#### Inference Stage (`DFPInferenceStage`)
 The `DFPInferenceStage` [examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_inference_stage.py](/examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_inference_stage.py) stage loads models from MLflow and performs inferences against those models.  This stage emits a message containing the original `DataFrame` along with new columns containing the z score (`mean_abs_z`), along with the name and version of the model that generated that score (`model_version`).  For each feature in the model three additional columns will also be added:
```
Comment (efajardo-nv) [inline docs/source/developer_guide/guides/5_digital_fingerprinting.md:566]: Add paranthesis to link to be consistent with other links or make class name a link.

File: docs/source/developer_guide/guides/5_digital_fingerprinting.md [hunk 11]
```diff
@@ -290,48 +537,54 @@ After training the generic model, individual user models can be trained​.  Ind
...
 * `<feature name>_loss` : The loss
 * `<feature name>_z_loss` : The loss z-score
 * `<feature name>_pred` : The predicted value
```
Comment (efajardo-nv) [inline docs/source/developer_guide/guides/5_digital_fingerprinting.md:569]: Not sure if this just a difference with the production pipeline or if something changed but starter dfp returns just `<feature name>_z_loss`. The other result columns are `reconstruct_loss`, `zscore`, 'max_abs_z', and 'mean_abs_z' which are not specific to a feature.

File: docs/source/developer_guide/guides/5_digital_fingerprinting.md [hunk 12]
```diff
@@ -290,48 +537,54 @@ After training the generic model, individual user models can be trained​.  Ind
...
+| Argument | Type | Description |
 | -------- | ---- | ----------- |
 | `c` | `morpheus.config.Config` | Morpheus config object |
-| `model_name_formatter` | `str` | Format string to control the name of models fetched from MLflow.  Currently available field names are: `user_id`. |
+| `model_name_formatter` | `str` | Format string to control the name of models fetched from MLflow.  Currently available field names are: `user_id` and `user_md5` which is an md5 hexadecimal digest as returned by [`hash.hexdigest`](https://docs.python.org/3.8/library/hashlib.html?highlight=hexdigest#hashlib.hash.hexdigest). |
+
 
 
-#### DFPPostprocessingStage
-The `DFPPostprocessingStage` [examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_postprocessing_stage.py](/examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_postprocessing_stage.py) stage filters records contained in the `DataFrame` emits a new `dfp.messages.DFPMessageMeta` message containing all records with a z score greater than `z_score_threshold` or `None` if no records are greater than `z_score_threshold`.
```
Comment (efajardo-nv) [inline docs/source/developer_guide/guides/5_digital_fingerprinting.md:332]: Add paranthesis to link to be consistent with other links or make class name a link.

File: docs/source/developer_guide/guides/5_digital_fingerprinting.md (hunk 11 above)
Comment (dagardner-nv) [reply docs/source/developer_guide/guides/5_digital_fingerprinting.md:569]: OK, this is a bit outside of my area, but the inference stage in the production example makes a call to `AutoEncoder.get_results`

which I believe is what this bit of code from that function is doing:
```
        for i, ft in enumerate(self.numeric_fts):
            pdf[ft] = df[ft]
            pdf[ft + '_pred'] = output_df[ft]
            pdf[ft + '_loss'] = mse[:, i].cpu().numpy()
            pdf[ft + '_z_loss'] = mse_scaled[:, i].cpu().numpy()

        for i, ft in enumerate(self.binary_fts):
            pdf[ft] = df[ft]
            pdf[ft + '_pred'] = output_df[ft]
            pdf[ft + '_loss'] = bce[:, i].cpu().numpy()
            pdf[ft + '_z_loss'] = bce_scaled[:, i].cpu().numpy()

        for i, ft in enumerate(self.categorical_fts):
            pdf[ft] = df[ft]
            pdf[ft + '_pred'] = output_df[ft]
            pdf[ft + '_loss'] = cce[:, i].cpu().numpy()
            pdf[ft + '_z_loss'] = cce_scaled[:, i].cpu().numpy()

        pdf['max_abs_z'] = combined_loss.max(dim=1)[0].cpu().numpy()
        pdf['mean_abs_z'] = combined_loss.mean(dim=1).cpu().numpy()
```

**Comment by dagardner-nv** [comment]
rerun tests

**Review by efajardo-nv — APPROVED** [review]

File: examples/digital_fingerprinting/production/morpheus/dfp/stages/multi_file_source.py [hunk 13]
```diff
@@ -69,7 +69,7 @@ def supports_cpp_node(self):
 
     def _generate_frames_fsspec(self):
 
-        files: fsspec.core.OpenFiles = fsspec.open_files(self._filenames, filecache={'cache_storage': './.cache/s3tmp'})
+        files: fsspec.core.OpenFiles = fsspec.open_files(self._filenames)
```
Comment (drobison00) [inline examples/digital_fingerprinting/production/morpheus/dfp/stages/multi_file_source.py:72]: Does caching not work correctly, or did we just decide not to do it?

**Review by drobison00 — APPROVED** [review]

File: examples/digital_fingerprinting/production/mlflow/Dockerfile [hunk 14]
```diff
@@ -24,7 +24,7 @@ RUN apt update && \
     rm -rf /var/cache/apt/* /var/lib/apt/lists/*
 
 # Install python packages
-RUN pip install mlflow boto3 pymysql pyyaml
+RUN pip install "mlflow<1.29.0" boto3 pymysql pyyaml
```
[the remark this answers is not in the document]
Comment (dagardner-nv) [reply examples/digital_fingerprinting/production/mlflow/Dockerfile:27]: @pdmack found a bug in that version and suggested the change.

File: examples/digital_fingerprinting/production/morpheus/dfp/stages/multi_file_source.py (hunk 13 above)
Comment (dagardner-nv) [reply examples/digital_fingerprinting/production/morpheus/dfp/stages/multi_file_source.py:72]: Leaving this in causes a keyword argument error when using an s3 source (works for local file sources). The `filecache` arg isn't documented. 
Users wanting to use a filecache can still do so by prefixing their file urls with `filecache::`

**Comment by dagardner-nv** [comment]
rerun tests

**Review by drobison00 — APPROVED** [review]

**Comment by dagardner-nv** [comment]
\rerun tests

**Comment by drobison00** [comment]
@gpucibot merge

**Pull request merged**